### PR TITLE
Feature 2027 grid stat wind direction

### LINKED ISSUE #2027: Investigate nightly build output wind direction differences caused by machine precision.
Note that this issue originally indicated that issues in Grid-Stat and Madis2NC should be investigated. However, the Madis2NC details were moved to a separate issue #2028 that @hsoh-u is investigating.

There are differences in the output generated from the MET-develop-ref data between machines. This is likely due to something different with an external dependency.

I copied data from the nightly builds from seneca and kiowa into a directory on seneca and ran the comp_dir.sh script.

Files are under:
/d1/personal/mccabe/kiowa_vs_seneca

> comp_dir_grid_stat.log
> comp_dir_madis2nc.log

These log files were generated by running the following commands:

`~mccabe/MET/test/bin/comp_dir.sh /d1/personal/mccabe/kiowa_vs_seneca/kiowa/NB20220125/MET-develop-ref/test_output/grid_stat /d1/personal/mccabe/kiowa_vs_seneca/seneca/NB20220125/MET-develop-ref/test_output/grid_stat &> comp_dir_grid_stat.log`

> 
> COMPARING grid_stat_GFS_FOURIER_240000L_20120410_000000V_pairs.nc
> file1: /d1/personal/mccabe/kiowa_vs_seneca/kiowa/NB20220125/MET-develop-ref/test_output/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V_pairs.nc
> file2: /d1/personal/mccabe/kiowa_vs_seneca/seneca/NB20220125/MET-develop-ref/test_output/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V_pairs.nc
> passed numerical var CLIMO_MEAN_HGT_P500_FULL
> passed numerical var CLIMO_MEAN_HGT_P500_FULL_WV1_0-3.
> ..
> passed numerical var FCST_HGT_P500_FULL
> passed numerical var FCST_HGT_P500_FULL_WV1_0-3
> **ERROR: found     2 differences in var FCST_HGT_P500_FULL_WV1_0-72            - max abs: 0.0004882812**
> passed numerical var FCST_HGT_P500_FULL_WV1_10-20
> ...
> 
> COMPARING grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
> file1: /d1/personal/mccabe/kiowa_vs_seneca/kiowa/NB20220125/MET-develop-ref/test_output/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
> file2: /d1/personal/mccabe/kiowa_vs_seneca/seneca/NB20220125/MET-develop-ref/test_output/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
> ERROR: found     8 differences in row type CNT          column FBAR          - max abs: 0.00000000000000008326
> ERROR: found     8 differences in row type CNT          column OBAR          - max abs: 0.00000000000000002775
> ERROR: found     8 differences in row type CNT          column ME            - max abs: 0.000000000000000013115
> ERROR: found     8 differences in row type CNT          column ME2           - max abs: 0.0000000000000000000000000000000022368
> ERROR: found     8 differences in row type SAL1L2       column FABAR         - max abs: 0.00000000000000006939
> ERROR: found     8 differences in row type SAL1L2       column OABAR         - max abs: 0.0000000000000000763
> ERROR: found     8 differences in row type SL1L2        column FBAR          - max abs: 0.00000000000000008326
> ERROR: found     8 differences in row type SL1L2        column OBAR          - max abs: 0.00000000000000002775
> ERROR: found     2 differences in row type VAL1L2       column UFABAR        - max abs: 0.0000000000000000001999
> ERROR: found     2 differences in row type VAL1L2       column VFABAR        - max abs: 0.000000000000000003469
> ERROR: found     2 differences in row type VAL1L2       column UOABAR        - max abs: 0.000000000000000001735
> ERROR: found     2 differences in row type VAL1L2       column VOABAR        - max abs: 0.00000000000000000165
> ERROR: found     2 differences in row type VCNT         column FBAR_SPEED    - max abs: 0.000000000000000004651
> ERROR: found     2 differences in row type VCNT         column OBAR_SPEED    - max abs: 0.000000000000000001861
> ERROR: found     2 differences in row type VCNT         column VDIFF_SPEED   - max abs: 0.000000000000000003135
> ERROR: found     2 differences in row type VCNT         column SPEED_ERR     - max abs: 0.000000000000000002968
> ERROR: found     2 differences in row type VCNT         column SPEED_ABSERR  - max abs: 0.000000000000000002968
> **ERROR: found     2 differences in row type VCNT         column DIR_ERR       - max abs: 2.3507
> ERROR: found     2 differences in row type VCNT         column DIR_ABSERR    - max abs: 2.3507**
> ERROR: found     2 differences in row type VL1L2        column UFBAR         - max abs: 0.00000000000000000094864
> ERROR: found     2 differences in row type VL1L2        column VFBAR         - max abs: 0.000000000000000005204
> ERROR: found     2 differences in row type VL1L2        column UOBAR         - max abs: 0.000000000000000000867
> ERROR: found     2 differences in row type VL1L2        column VOBAR         - max abs: 0.000000000000000001952
> 
> SUMMARY for non-bootstrap numerical values
> \# comparisons: 1575
> \# differences: 94
>        Min.: -2.3507
>     1st Qu.: -1.844e-18
>      Median: -2.4458e-19
>        Mean: -0.05001489
>     3rd Qu.: 1.2587e-18
>        Max.: 1.48229
> 
> **ERROR: failed tests for CNT: num
> ERROR: failed tests for SAL1L2: num
> ERROR: failed tests for SL1L2: num
> ERROR: failed tests for VAL1L2: num
> ERROR: failed tests for VCNT: num
> ERROR: failed tests for VL1L2: num**

`~mccabe/MET/test/bin/comp_dir.sh /d1/personal/mccabe/kiowa_vs_seneca/kiowa/NB20220125/MET-develop-ref/test_output/madis2nc /d1/personal/mccabe/kiowa_vs_seneca/seneca/NB20220125/MET-develop-ref/test_output/madis2nc &> comp_dir_madis2nc.log`

>
> COMPARING profiler_2012040912_F000.nc
> file1: /d1/personal/mccabe/kiowa_vs_seneca/kiowa/NB20220125/MET-develop-ref/test_output/madis2nc/profiler_2012040912_F000.nc
> file2: /d1/personal/mccabe/kiowa_vs_seneca/seneca/NB20220125/MET-develop-ref/test_output/madis2nc/profiler_2012040912_F000.nc
> ERROR: NetCDF headers differ:
> 11c11
> \<       nobs_qty = 2 ;
> ---
> \>       nobs_qty = 1 ;

## Describe the Task ##
Investigate these differences and create a GitHub issue if a fix to the source code would prevent these differences. If changes to the external libraries used to build MET on seneca resolve these differences, then document those changes on this issue.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by hsoh-u**
## Expected Differences ##

There are minor differences because of precision (different hardware and different compiler versions)

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

  - Run unit test at kiowa and seneca
  - Compare unit test output (especially grid_stat)

The unit test outputs are at
 -  kiowa:/d1/personal/hsoh/git/pull_request/MET_feature_2027_grid_stat_wind_direction/unit_test_output/
 -  seneca:/d1/personal/hsoh/git/features/feature_2027_grid_stat_wind_direction/MET/unit_test_output/

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [x] Do these changes include sufficient testing updates? **[No]**

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

The threshold for wind is changed to ">5.0". The output of grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat is changed.

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/libcode/vx_statistics/met_stats.cc [hunk 1]
```diff
@@ -1569,6 +1570,12 @@ void VL1L2Info::calc_ncep_stats() {
 
    DIR_ABSERR   = fabs(DIR_ERR);
 
+   char buffer[1024];
```
Comment (JohnHalleyGotway) [inline met/src/libcode/vx_statistics/met_stats.cc]: @hsoh-u this looks like temporary code that you added to debug this issue. But we wouldn't want to actually include this in the release. Recommend removing.

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
Howard, I'd recommend removing that debug code from met_stats.cc.

If you see some need to keep it, I'd recommend replacing the use of that static character buffer with a call to ConcatString::format() instead.

**Comment by hsoh-u** [comment]
It's removed

**Review by JohnHalleyGotway — APPROVED** [review]
Simple change. Looks good to me. Thanks Howard.

Note that this will change the output of the unit tests and we'll need to merge develop into develop-ref after merging this into develop.

All the unit tests ran, but as expected, differences were flagged in the output. The diffs are limited to vector partial sums and statistics in a single Grid-Stat output file, which is exactly what they should be based on the change to the Grid-Stat config file.

```
COMPARING grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
file1: /data/output/met_test_truth/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
file2: /data/output/met_test_output/grid_stat/grid_stat_GFS_FOURIER_240000L_20120410_000000V.stat
```

**Pull request merged**